File: packages/protocol/src/transfer-sender.test.ts
import { describe, it, expect } from 'vitest';
import { createHash } from 'node:crypto';
import { deriveTransferKeys } from './keyschedule.js';
import { open, seal } from './aead.js';
import { FrameType } from './transfer.js';
import { FRAME_VERSION } from './constants.js';
import { decodeControl, encodeControl } from './transfer-messages.js';
import { bytesSource, type Digest } from './transfer-ports.js';
import { TransferSender } from './transfer-sender.js';

function nodeDigest(): Digest {
  const h = createHash('sha256');
  return { update: (b) => void h.update(b), hexDigest: () => h.digest('hex') };
}
const master = new Uint8Array(32).fill(9);

async function waitFor(check: () => boolean): Promise<void> {
  const deadline = Date.now() + 1000;
  while (!check()) {
    if (Date.now() >= deadline) throw new Error('timed out waiting for sender state');
    await new Promise((resolve) => setTimeout(resolve, 1));
  }
}

describe('TransferSender', () => {
  it('emits manifest → block_data/block_hash → complete, gated by verified acknowledgements', async () => {
    const keys = await deriveTransferKeys(master);
    const data = new Uint8Array(20).map((_, i) => i); // block=8 → blocks 0,1 full, block2 = 4B
    const outbound: Uint8Array[] = [];
    const progress: number[] = [];

    const sender = new TransferSender({
      file: bytesSource(data, { name: 'f', size: 20, mime: '', lastModified: 0 }),
      send: (f) => void outbound.push(f),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize: 8,
      frameSize: 4,
      window: 1, // force gating after block 0
      onProgress: (bytes) => progress.push(bytes),
    });

    const runP = sender.run();

    // Let the sender push manifest + block 0's frames + block 0 hash, then stall on window=1.
    await new Promise((r) => setTimeout(r, 10));
    let peek = 0;
    const types = await Promise.all(
      outbound.map(async (f) => (await open(keys.o2j, peek++, f)).header.type),
    );
    expect(types[0]).toBe(FrameType.Manifest);
    // With window=1 the sender may run 1 block ahead; it must NOT have sent `complete` yet.
    expect(types).not.toContain(FrameType.Complete);

    // Feed a verified acknowledgement for each block so the window drains, then `done`.
    let ackCtr = 0;
    const ack = async (blockIdx: number) => {
      const frame = await seal(
        keys.j2o,
        ackCtr++,
        {
          version: FRAME_VERSION,
          type: FrameType.Ack,
          flags: 0,
          fileIdx: 0,
          blockIdx: 0,
          frameOff: 0,
        },
        encodeControl({ type: FrameType.Ack, fileIdx: 0, blockIdx }),
      );
      sender.handle(frame);
    };
    await ack(0);
    await new Promise((r) => setTimeout(r, 10));
    await ack(1);
    await new Promise((r) => setTimeout(r, 10));
    await ack(2);
    await new Promise((r) => setTimeout(r, 10));
    const done = await seal(
      keys.j2o,
      ackCtr++,
      {
        version: FRAME_VERSION,
        type: FrameType.Done,
        flags: 0,
        fileIdx: 0,
        blockIdx: 0,
        frameOff: 0,
      },
      encodeControl({ type: FrameType.Done }),
    );
    sender.handle(done);
    await runP;
    expect(progress).toEqual([8, 16, 20]);

    // Decode the full outbound sequence and check the message grammar + digest.
    let c = 0;
    const decoded: Array<{ type: number; payload: Uint8Array }> = [];
    for (const f of outbound) {
      const o = await open(keys.o2j, c++, f);
      decoded.push({ type: o.header.type, payload: o.plaintext });
    }
    expect(decoded.at(-1)!.type).toBe(FrameType.Complete);
    const complete = decodeControl(decoded.at(-1)!.payload);
    if (complete.type !== FrameType.Complete) throw new Error('expected complete');
    expect(complete.fileDigest).toBe(createHash('sha256').update(data).digest('hex'));
    // Reassemble block_data payloads → original bytes.
    const body: number[] = [];
    for (const d of decoded) if (d.type === FrameType.BlockData) body.push(...d.payload);
    expect(body).toEqual([...data]);
  });

  it('rejects run() when the receiver reports fail', async () => {
    const keys = await deriveTransferKeys(master);
    const sender = new TransferSender({
      file: bytesSource(new Uint8Array(4), { name: 'f', size: 4, mime: '', lastModified: 0 }),
      send: () => {},
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize: 8,
      frameSize: 4,
    });
    const runP = sender.run();
    await new Promise((r) => setTimeout(r, 0));
    const fail = await seal(
      keys.j2o,
      0,
      {
        version: FRAME_VERSION,
        type: FrameType.Fail,
        flags: 0,
        fileIdx: 0,
        blockIdx: 0,
        frameOff: 0,
      },
      encodeControl({ type: FrameType.Fail, reason: 'integrity' }),
    );
    sender.handle(fail);
    await expect(runP).rejects.toThrow(/integrity/);
  });

  it('reseals a nacked block with fresh counters and advances progress only on ack', async () => {
    const keys = await deriveTransferKeys(master);
    const data = new Uint8Array([1, 2, 3, 4]);
    const outbound: Uint8Array[] = [];
    const progress: number[] = [];
    const sender = new TransferSender({
      file: bytesSource(data, { name: 'f', size: data.length, mime: '', lastModified: 0 }),
      send: (frame) => void outbound.push(frame),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize: 8,
      frameSize: 4,
      window: 1,
      ackTimeoutMs: 1000,
      onProgress: (bytes) => progress.push(bytes),
    });

    const runP = sender.run();
    await waitFor(() => outbound.length === 3); // manifest, block_data, block_hash
    expect(progress).toEqual([]);

    const nack = await seal(
      keys.j2o,
      0,
      {
        version: FRAME_VERSION,
        type: FrameType.Nack,
        flags: 0,
        fileIdx: 0,
        blockIdx: 0,
        frameOff: 0,
      },
      encodeControl({ type: FrameType.Nack, fileIdx: 0, blockIdx: 0, reason: 'missing' }),
    );
    sender.handle(nack);
    await waitFor(() => outbound.length === 5); // freshly sealed data + hash

    const ack = await seal(
      keys.j2o,
      1,
      {
        version: FRAME_VERSION,
        type: FrameType.Ack,
        flags: 0,
        fileIdx: 0,
        blockIdx: 0,
        frameOff: 0,
      },
      encodeControl({ type: FrameType.Ack, fileIdx: 0, blockIdx: 0 }),
    );
    sender.handle(ack);
    await waitFor(() => outbound.length === 6); // complete
    expect(progress).toEqual([4]);

    const done = await seal(
      keys.j2o,
      2,
      {
        version: FRAME_VERSION,
        type: FrameType.Done,
        flags: 0,
        fileIdx: 0,
        blockIdx: 0,
        frameOff: 0,
      },
      encodeControl({ type: FrameType.Done }),
    );
    sender.handle(done);
    await runP;

    let counter = 0;
    const opened = [];
    for (const frame of outbound) opened.push(await open(keys.o2j, counter++, frame));
    expect(opened.map((item) => item.header.type)).toEqual([
      FrameType.Manifest,
      FrameType.BlockData,
      FrameType.BlockHash,
      FrameType.BlockData,
      FrameType.BlockHash,
      FrameType.Complete,
    ]);
    expect(opened[1]?.plaintext).toEqual(data);
    expect(opened[3]?.plaintext).toEqual(data);
    expect(outbound[1]).not.toEqual(outbound[3]); // retransmission uses a fresh GCM nonce
  });

  it('immediately retransmits the unacknowledged window after a transport change', async () => {
    const keys = await deriveTransferKeys(master);
    const data = new Uint8Array([1, 2, 3, 4]);
    const outbound: Uint8Array[] = [];
    const sender = new TransferSender({
      file: bytesSource(data, { name: 'f', size: data.length, mime: '', lastModified: 0 }),
      send: (frame) => void outbound.push(frame),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize: 8,
      frameSize: 4,
      window: 1,
      ackTimeoutMs: 60_000,
    });

    const run = sender.run();
    await waitFor(() => outbound.length === 3);
    sender.transportChanged();
    await waitFor(() => outbound.length === 5);
    sender.cancel('test complete');
    await expect(run).rejects.toThrow(/test complete/);

    let counter = 0;
    const opened = [];
    for (const frame of outbound.slice(0, 5)) opened.push(await open(keys.o2j, counter++, frame));
    expect(opened.map((frame) => frame.header.type)).toEqual([
      FrameType.Manifest,
      FrameType.BlockData,
      FrameType.BlockHash,
      FrameType.BlockData,
      FrameType.BlockHash,
    ]);
    expect(outbound[1]).not.toEqual(outbound[3]);
  });

  it('halts new data while paused and resumes without losing the window', async () => {
    const keys = await deriveTransferKeys(master);
    const outbound: Uint8Array[] = [];
    const states: string[] = [];
    const sender = new TransferSender({
      file: bytesSource(new Uint8Array([1, 2, 3, 4]), {
        name: 'f',
        size: 4,
        mime: '',
        lastModified: 0,
      }),
      send: (frame) => void outbound.push(frame),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize: 8,
      frameSize: 4,
      onStateChange: (state) => states.push(state),
    });

    sender.pause();
    const runP = sender.run();
    await waitFor(() => outbound.length === 2); // pause control + manifest
    await new Promise((resolve) => setTimeout(resolve, 10));
    expect(outbound).toHaveLength(2);

    sender.resume();
    await waitFor(() => outbound.length === 5); // resume + block_data + block_hash
    const ack = await seal(
      keys.j2o,
      0,
      {
        version: FRAME_VERSION,
        type: FrameType.Ack,
        flags: 0,
        fileIdx: 0,
        blockIdx: 0,
        frameOff: 0,
      },
      encodeControl({ type: FrameType.Ack, fileIdx: 0, blockIdx: 0 }),
    );
    sender.handle(ack);
    await waitFor(() => outbound.length === 6);
    const done = await seal(
      keys.j2o,
      1,
      {
        version: FRAME_VERSION,
        type: FrameType.Done,
        flags: 0,
        fileIdx: 0,
        blockIdx: 0,
        frameOff: 0,
      },
      encodeControl({ type: FrameType.Done }),
    );
    sender.handle(done);
    await runP;
    expect(states).toEqual(['paused', 'running']);
  });
});

File: packages/wire/transfer_sender_test.go
package wire

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// senderMaster is the fixed 32-byte master key both sender tests derive their keys from.
func senderMaster() []byte {
	m := make([]byte, 32)
	for i := range m {
		m[i] = 9
	}
	return m
}

// outbox is a concurrency-safe capture of the frames a Sender emits from its Run goroutine.
type outbox struct {
	mu     sync.Mutex
	frames [][]byte
}

func (o *outbox) push(f []byte) error {
	o.mu.Lock()
	o.frames = append(o.frames, append([]byte(nil), f...))
	o.mu.Unlock()
	return nil
}

func (o *outbox) len() int {
	o.mu.Lock()
	defer o.mu.Unlock()
	return len(o.frames)
}

func (o *outbox) snapshot() [][]byte {
	o.mu.Lock()
	defer o.mu.Unlock()
	return append([][]byte(nil), o.frames...)
}

// waitStable blocks until the outbox has held want frames unchanged across a few polls,
// i.e. the sender has stalled on its window gate rather than merely being between sends.
func waitStable(t *testing.T, o *outbox, want int) {
	t.Helper()
	stable := 0
	for i := 0; i < 2000; i++ {
		if o.len() == want {
			if stable++; stable >= 5 {
				return
			}
		} else {
			stable = 0
		}
		time.Sleep(time.Millisecond)
	}
	t.Fatalf("outbox never stabilized at %d frames (saw %d)", want, o.len())
}

func TestSenderEmitsGrammarGatedByAck(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	data := seq(20) // block=8 → blocks 0,1 full, block 2 = 4 bytes
	var out outbox

	s := NewSender(SenderOptions{
		File:      BytesSource(data, FileMeta{Name: "f", Size: 20}, 0),
		Send:      out.push,
		SendDir:   keys.O2J,
		RecvDir:   keys.J2O,
		BlockSize: 8,
		FrameSize: 4,
		Window:    1, // force gating after block 0
	})

	type result struct {
		digest string
		err    error
	}
	res := make(chan result, 1)
	go func() {
		d, err := s.Run(context.Background())
		res <- result{d, err}
	}()

	// manifest + block 0's two frames + block 0 hash = 4 frames, then it stalls on window=1.
	waitStable(t, &out, 4)
	peek := 0
	for _, f := range out.snapshot() {
		o, err := Open(keys.O2J, uint64(peek), f)
		if err != nil {
			t.Fatalf("open frame %d: %v", peek, err)
		}
		if peek == 0 && o.Header.Type != FrameManifest {
			t.Fatalf("frame 0 type = %d, want manifest", o.Header.Type)
		}
		if o.Header.Type == FrameComplete {
			t.Fatalf("complete sent before receiver acked the window")
		}
		peek++
	}

	// Ack every verified block so the window drains, then send done.
	ackCtr := uint64(0)
	ack := func(blockIdx int) {
		payload, err := EncodeControl(NewAck(0, blockIdx))
		if err != nil {
			t.Fatal(err)
		}
		frame, err := Seal(keys.J2O, ackCtr, FrameHeaderInput{Version: FrameVersion, Type: FrameAck}, payload)
		if err != nil {
			t.Fatal(err)
		}
		ackCtr++
		s.Handle(frame)
	}
	ack(0)
	waitStable(t, &out, 7)
	ack(1)
	waitStable(t, &out, 9)
	ack(2)

	// Every block, including the short final block, carries a hash. The sender emits complete
	// only after all three verified acknowledgements arrive.
	waitStable(t, &out, 10)

	donePayload, err := EncodeControl(NewDone())
	if err != nil {
		t.Fatal(err)
	}
	doneFrame, err := Seal(keys.J2O, ackCtr, FrameHeaderInput{Version: FrameVersion, Type: FrameDone}, donePayload)
	if err != nil {
		t.Fatal(err)
	}
	s.Handle(doneFrame)

	got := <-res
	if got.err != nil {
		t.Fatalf("Run: %v", got.err)
	}
	sum := sha256.Sum256(data)
	wantDigest := hex.EncodeToString(sum[:])
	if got.digest != wantDigest {
		t.Errorf("digest = %s, want %s", got.digest, wantDigest)
	}

	// Decode the full outbound sequence: last frame must be complete with the file digest, and
	// the block_data payloads must reassemble to the original bytes in order.
	var body []byte
	var lastType uint8
	var lastPayload []byte
	c := 0
	for _, f := range out.snapshot() {
		o, err := Open(keys.O2J, uint64(c), f)
		if err != nil {
			t.Fatalf("open frame %d: %v", c, err)
		}
		if o.Header.Type == FrameBlockData {
			body = append(body, o.Plaintext...)
		}
		lastType, lastPayload = o.Header.Type, o.Plaintext
		c++
	}
	if lastType != FrameComplete {
		t.Fatalf("last frame type = %d, want complete", lastType)
	}
	msg, err := DecodeControl(lastPayload)
	if err != nil {
		t.Fatal(err)
	}
	complete, ok := msg.(*Complete)
	if !ok || complete.FileDigest != wantDigest {
		t.Fatalf("complete = %#v, want digest %s", msg, wantDigest)
	}
	if string(body) != string(data) {
		t.Errorf("reassembled body = %v, want %v", body, data)
	}
}

func TestSenderRejectsOnReceiverFail(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	s := NewSender(SenderOptions{
		File:      BytesSource(make([]byte, 4), FileMeta{Name: "f", Size: 4}, 0),
		Send:      func([]byte) error { return nil },
		SendDir:   keys.O2J,
		RecvDir:   keys.J2O,
		BlockSize: 8,
		FrameSize: 4,
	})

	res := make(chan error, 1)
	go func() {
		_, err := s.Run(context.Background())
		res <- err
	}()

	payload, err := EncodeControl(NewFail(FailIntegrity))
	if err != nil {
		t.Fatal(err)
	}
	failFrame, err := Seal(keys.J2O, 0, FrameHeaderInput{Version: FrameVersion, Type: FrameFail}, payload)
	if err != nil {
		t.Fatal(err)
	}
	s.Handle(failFrame)

	select {
	case err := <-res:
		if err == nil || !strings.Contains(err.Error(), "integrity") {
			t.Fatalf("Run error = %v, want one mentioning integrity", err)
		}
	case <-time.After(2 * time.Second):
		t.Fatal("Run did not return after receiver fail")
	}
}

func TestSenderRetransmitsNackWithFreshCounterAndAckedProgress(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	data := []byte{1, 2, 3, 4}
	var out outbox
	var progress []int64
	s := NewSender(SenderOptions{
		File:       BytesSource(data, FileMeta{Name: "f", Size: int64(len(data))}, 0),
		Send:       out.push,
		SendDir:    keys.O2J,
		RecvDir:    keys.J2O,
		BlockSize:  8,
		FrameSize:  4,
		Window:     1,
		AckTimeout: time.Second,
		OnProgress: func(n int64) { progress = append(progress, n) },
	})

	res := make(chan error, 1)
	go func() { _, runErr := s.Run(context.Background()); res <- runErr }()
	waitStable(t, &out, 3)
	if len(progress) != 0 {
		t.Fatalf("progress before ack = %v", progress)
	}

	peerCounter := uint64(0)
	feed := func(msg ControlMsg) {
		payload, encodeErr := EncodeControl(msg)
		if encodeErr != nil {
			t.Fatal(encodeErr)
		}
		frame, sealErr := Seal(keys.J2O, peerCounter,
			FrameHeaderInput{Version: FrameVersion, Type: msg.FrameType()}, payload)
		if sealErr != nil {
			t.Fatal(sealErr)
		}
		peerCounter++
		s.Handle(frame)
	}
	feed(NewNack(0, 0, NackMissing))
	waitStable(t, &out, 5)
	feed(NewAck(0, 0))
	waitStable(t, &out, 6)
	feed(NewDone())
	if runErr := <-res; runErr != nil {
		t.Fatalf("Run: %v", runErr)
	}
	if len(progress) != 1 || progress[0] != int64(len(data)) {
		t.Fatalf("progress = %v, want [%d]", progress, len(data))
	}

	frames := out.snapshot()
	wantTypes := []uint8{FrameManifest, FrameBlockData, FrameBlockHash,
		FrameBlockData, FrameBlockHash, FrameComplete}
	for i, frame := range frames {
		opened, openErr := Open(keys.O2J, uint64(i), frame)
		if openErr != nil {
			t.Fatalf("open frame %d: %v", i, openErr)
		}
		if opened.Header.Type != wantTypes[i] {
			t.Errorf("frame %d type = %d, want %d", i, opened.Header.Type, wantTypes[i])
		}
	}
	if string(frames[1]) == string(frames[3]) {
		t.Error("retransmitted ciphertext reused the original nonce")
	}
}

func TestSenderPauseResumeStopsNewData(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	var out outbox
	var states []TransferState
	s := NewSender(SenderOptions{
		File:          BytesSource([]byte{1, 2, 3, 4}, FileMeta{Name: "f", Size: 4}, 0),
		Send:          out.push,
		SendDir:       keys.O2J,
		RecvDir:       keys.J2O,
		BlockSize:     8,
		FrameSize:     4,
		OnStateChange: func(state TransferState) { states = append(states, state) },
	})
	if err := s.Pause(); err != nil {
		t.Fatal(err)
	}
	res := make(chan error, 1)
	go func() { _, runErr := s.Run(context.Background()); res <- runErr }()
	waitStable(t, &out, 2) // pause control + manifest
	if err := s.Resume(); err != nil {
		t.Fatal(err)
	}
	waitStable(t, &out, 5) // resume control + data + hash

	peerCounter := uint64(0)
	for _, msg := range []ControlMsg{NewAck(0, 0), NewDone()} {
		if _, ok := msg.(*Done); ok {
			waitStable(t, &out, 6) // complete precedes done
		}
		payload, encodeErr := EncodeControl(msg)
		if encodeErr != nil {
			t.Fatal(encodeErr)
		}
		frame, sealErr := Seal(keys.J2O, peerCounter,
			FrameHeaderInput{Version: FrameVersion, Type: msg.FrameType()}, payload)
		if sealErr != nil {
			t.Fatal(sealErr)
		}
		peerCounter++
		s.Handle(frame)
	}
	if runErr := <-res; runErr != nil {
		t.Fatalf("Run: %v", runErr)
	}
	if len(states) != 2 || states[0] != TransferPaused || states[1] != TransferRunning {
		t.Fatalf("states = %v, want [paused running]", states)
	}
}

func TestSenderBoundsTimeoutRetries(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	var out outbox
	s := NewSender(SenderOptions{
		File:       BytesSource([]byte{1, 2, 3, 4}, FileMeta{Name: "f", Size: 4}, 0),
		Send:       out.push,
		SendDir:    keys.O2J,
		RecvDir:    keys.J2O,
		BlockSize:  8,
		FrameSize:  4,
		AckTimeout: 5 * time.Millisecond,
		MaxRetries: 1,
	})
	res := make(chan error, 1)
	go func() { _, runErr := s.Run(context.Background()); res <- runErr }()
	select {
	case runErr := <-res:
		if runErr == nil || !strings.Contains(runErr.Error(), string(FailRetryExhausted)) {
			t.Fatalf("Run error = %v, want retry_exhausted", runErr)
		}
	case <-time.After(2 * time.Second):
		t.Fatal("Run did not stop after bounded retries")
	}
	if out.len() != 6 { // manifest + initial data/hash + one retry data/hash + fail
		t.Fatalf("outbox has %d frames, want 6", out.len())
	}
}

func TestSenderContextCancellationNotifiesPeer(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	var out outbox
	s := NewSender(SenderOptions{
		File: BytesSource([]byte{1, 2, 3, 4}, FileMeta{Name: "f", Size: 4}, 0),
		Send: out.push, SendDir: keys.O2J, RecvDir: keys.J2O,
		BlockSize: 8, FrameSize: 4, AckTimeout: time.Second,
	})
	ctx, cancel := context.WithCancel(context.Background())
	res := make(chan error, 1)
	go func() { _, runErr := s.Run(ctx); res <- runErr }()
	waitStable(t, &out, 3)
	cancel()
	select {
	case runErr := <-res:
		if runErr == nil || !strings.Contains(runErr.Error(), string(FailCanceled)) {
			t.Fatalf("Run error = %v, want canceled", runErr)
		}
	case <-time.After(time.Second):
		t.Fatal("Run did not stop after context cancellation")
	}
	frames := out.snapshot()
	opened, openErr := Open(keys.O2J, uint64(len(frames)-1), frames[len(frames)-1])
	if openErr != nil {
		t.Fatal(openErr)
	}
	msg, decodeErr := DecodeControl(opened.Plaintext)
	control, ok := msg.(*Control)
	if decodeErr != nil || !ok || control.Op != ControlCancel {
		t.Fatalf("last cancellation frame = %#v, err=%v", msg, decodeErr)
	}
}

// TestSenderFailsOnDoneBeforeComplete pins the fail-closed half of the Done path: a
// receiver may only send Done in response to Complete (after verifying the whole-file
// digest), so Done before Complete is a protocol violation and must fail the transfer.
func TestSenderFailsOnDoneBeforeComplete(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	var out outbox
	s := NewSender(SenderOptions{
		File:      BytesSource([]byte{1, 2, 3, 4}, FileMeta{Name: "f", Size: 4}, 0),
		Send:      out.push,
		SendDir:   keys.O2J,
		RecvDir:   keys.J2O,
		BlockSize: 8,
		FrameSize: 4,
	})
	res := make(chan error, 1)
	go func() { _, runErr := s.Run(context.Background()); res <- runErr }()
	waitStable(t, &out, 3) // manifest + data + hash

	done, err := EncodeControl(NewDone())
	if err != nil {
		t.Fatal(err)
	}
	frame, err := Seal(keys.J2O, 0, FrameHeaderInput{Version: FrameVersion, Type: NewDone().FrameType()}, done)
	if err != nil {
		t.Fatal(err)
	}
	s.Handle(frame)
	if runErr := <-res; runErr == nil {
		t.Fatal("Run succeeded on Done before Complete was sent")
	}
}

func TestSenderOnManifestRunsBeforeFirstFrame(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	data := seq(20) // block=8 → blocks 0,1 full, block 2 = 4 bytes
	var out outbox
	hookDone := make(chan *Manifest, 1)
	var hookRan atomic.Bool
	firstFrameHookRan := false
	seenFirst := false
	s := NewSender(SenderOptions{
		File:       BytesSource(data, FileMeta{Name: "f", Size: 20}, 0),
		Send:       out.push,
		SendDir:    keys.O2J,
		RecvDir:    keys.J2O,
		BlockSize:  8,
		FrameSize:  4,
		Window:     1,
		TransferID: "0123456789abcdef0123456789abcdef",
		OnManifest: func(m Manifest) error {
			hookRan.Store(true)
			hookDone <- &m
			return nil
		},
	})
	// The hook must have completed before the very first frame (the manifest) is emitted:
	// the sender's own goroutine runs the hook, then Send; capturing order proves the
	// record would be durable before the id is advertised.
	origSend := out.push
	s.o.Send = func(f []byte) error {
		if !seenFirst {
			seenFirst = true
			firstFrameHookRan = hookRan.Load()
		}
		return origSend(f)
	}

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	res := make(chan error, 1)
	go func() { _, runErr := s.Run(ctx); res <- runErr }()

	select {
	case m := <-hookDone:
		if m.TransferID != "0123456789abcdef0123456789abcdef" {
			t.Fatalf("hook manifest transferId = %q", m.TransferID)
		}
		if len(m.Files) != 1 || m.Files[0].Name != "f" || m.Files[0].Size != 20 {
			t.Fatalf("hook manifest files = %#v", m.Files)
		}
		sum := sha256.Sum256(data)
		if m.Files[0].FileDigest != hex.EncodeToString(sum[:]) {
			t.Fatalf("hook manifest digest = %s", m.Files[0].FileDigest)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("OnManifest never called")
	}
	// The manifest frame must be on the wire before canceling: the cancellation sends its
	// own control frame, and racing it against the manifest send makes the "first frame is
	// the manifest" assertion below nondeterministic.
	waitStable(t, &out, 1)

	cancel()
	if err := <-res; err == nil {
		t.Fatalf("Run unexpectedly succeeded after cancellation")
	}
	if !seenFirst {
		t.Fatal("no frame was ever sent")
	}
	if !firstFrameHookRan {
		t.Fatal("the manifest frame was emitted before OnManifest completed")
	}
	if out.len() == 0 {
		t.Fatal("manifest frame missing after hook")
	}
	opened, err := Open(keys.O2J, 0, out.snapshot()[0])
	if err != nil {
		t.Fatalf("open first frame: %v", err)
	}
	if opened.Header.Type != FrameManifest {
		t.Fatalf("first frame type = %d, want manifest", opened.Header.Type)
	}
	msg, err := DecodeControl(opened.Plaintext)
	if err != nil {
		t.Fatal(err)
	}
	m, ok := msg.(*Manifest)
	if !ok || m.TransferID != "0123456789abcdef0123456789abcdef" {
		t.Fatalf("first frame manifest = %#v", msg)
	}
}

func TestSenderOnManifestErrorAbortsBeforeManifestSent(t *testing.T) {
	keys, err := DeriveTransferKeys(senderMaster())
	if err != nil {
		t.Fatal(err)
	}
	var out outbox
	s := NewSender(SenderOptions{
		File:       BytesSource(seq(20), FileMeta{Name: "f", Size: 20}, 0),
		Send:       out.push,
		SendDir:    keys.O2J,
		RecvDir:    keys.J2O,
		BlockSize:  8,
		FrameSize:  4,
		TransferID: "0123456789abcdef0123456789abcdef",
		OnManifest: func(Manifest) error {
			return errors.New("sender-state write failed")
		},
	})
	res := make(chan error, 1)
	go func() { _, runErr := s.Run(context.Background()); res <- runErr }()
	select {
	case runErr := <-res:
		if runErr == nil || !strings.Contains(runErr.Error(), "sender-state write failed") {
			t.Fatalf("Run error = %v, want the hook failure", runErr)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("Run did not return after hook failure")
	}
	if out.len() != 0 {
		t.Fatalf("hook failure still emitted %d frames; the manifest must never reach the wire", out.len())
	}
}
